### ISSUE TITLE
Weird space consumption after all threads deleted

### ISSUE THREAD

**Issue by sailbird** [signalapp/Signal-Android#10179]
- [x] I have searched open and closed issues for duplicates
- [x] I am submitting a bug report for existing functionality that does not work as intended
- [x] I have read https://github.com/signalapp/Signal-Android/wiki/Submitting-useful-bug-reports
- [x] This isn't a feature request or a discussion topic

----------------------------------------

### Bug description
Some context is necessary to explain why this matters. It's not so much about wasted space, as an attempt to guarantee that deleted messages are really unrecoverable, and what else we need to do to ensure this if no such guarantee can be made.

There are times, particularly when crossing from one territory into another, that people may be forced to unlock their phones for dumping by the authorities, so threads need to be deleted ahead of time and rendered permanently unrecoverable on the device. (A really smart border agent might force you to open all your devices, hoping to obtain duplicates of the deleted threads on your linked device, which is why devices need to be unlinked prior to the privacy invasion, but that's another matter.) It's critical that deleted threads never be recoverable even by direct dumping of the media, and even if the phone has been unlocked. On the other hand, it's equally critical that the association between a given Signal installation and the phone number used to register it remain intact across any process which destroys those deleted threads, mainly because it makes good security sense to register with a disposable SIM unrelated to the phone's actual SIM.

Now, the bug:

After building up hundreds of MB of chats including shared media, I deleted all my threads. None were archived. However, Signal's app info in the phone setup still shows that it's consuming enough space to still be storing them somewhere. I assume they're encrypted, but that's not necessarily good enough, which I'll get to below.

Of course, I could just clear the cache, but that would create another serious problem, which is that it would force me to reregister my Signal phone number. Why this has to occur just in order to clear the cache makes no sense to me; the two items just seemed to have been lumped together unnecessarily.

### Steps to reproduce

- Make a bunch of threads with messaging peers. They must be big enough so that they are many times the base install size of Signal. Therefore it will be obvious when the storage footprint is due to their associated allocations, as opposed to just Signal's executable code.

- Take note of Signal's storage footprint in the phone's app settings.

- Delete all the threads, one by one. You can even reboot the phone just to -- hopefully -- flush the deallocated space.

- Go back to the phone's app settings and look at Signal's footprint size once again.

**Actual result:**
The footprint has hardly changed. You can only change it by flushing Signal's cache, which leads to the problem mentioned above.

**Expected result:**
**Option 1**
Deleting messages, or entire message threads, doesn't free up space, but it deletes the unique-per-message crypto keys used to allow them to be decrypted, so anyone dumping the phone storage won't see any zeroed-out gaps between messages which might allow accurate message sizes to be determined. (I'm assuming some crude object store like: key hash 0, message 0, key hash 1, message 1, etc., out of total ignorance, but the point is that keys should not be zeroed out in order to delete messages; they should be filled with noise because while having one message size is probably useless, having a lot of them, even granularized, is potentially useful for various identification purposes.) To put it another way, I assume that each received message is locally stored and encrypted with a unique and randomly generated key, which just needs to be overwritten with noise upon deletion, so that the deleted message can never be recovered, even though it is still theoretically decryptable. (If this is not the case, then please explain what actually happens. Is there a per-thread key instead? A per-installation key? Or what? Per-thread would mean that we're only safe if we delete whole threads, for instance.) Now, in order to actually free the deallocated space, the user must visit the app's storage menu in the phone's settings: "Clear Cache" clears all data _except_ the association between the Signal installation and a particular phone number. This allows users to maintain that association long after the registering SIM has been trashed. "Clear Data" clears that assocation also, and potentially other user preferences, leaving a situation indistinct from a fresh install.

**Option 2**
"Clear Cache" kills the phone number association, so in order to avoid that, the user will need to delete one thread at a time. After deleting (deallocating) them, garbage collection happens at some point, hopefully before the phone is rebooted. In any event, after the reboot, all the space previously allocated to deleted threads has been relinquished to the phone's storage pool. ("Clear Data" performs as in Option 1.)

### Device info
**Device:** Samsung S20
**Android version:** 10
**Signal version:** (latest as of this report)

**Comment by greyson-signal** [comment]
It'd be useful if you were able to post before/after comparisons of what you see when you go to Settings > Storage so we can see how much data is left over and what type. Also, in Settings > Storage, there's a button to delete all message history. I'm curious if you get different results with that.

**Comment by sailbird** [comment]
@greyson-signal Sorry I missed your reply. I did some experiments and discovered that the internal notion of storage does not even close to match the OS notion. Go to Settings -> Apps -> Signal -> Storage and see the amount used by App, Data, and Cache. Then compare this to Signal Settings -> Data and storage -> Manage storage. The numbers are vastly different. We're not talking rounding error here.

I'm basing this off of what happened both before and after I clicked on "REVIEW STORAGE" and deleted a bunch of big messages. It wasn't accurate and still isn't. Curiously, though, the deletion caused both values to decrease by about the same amount, so that much might be in sync.

I would just assume delete Signal and reinstall, but that's a deep rabbit hole because then we get into all the security ramifications of reregistration. (I can't wait until you manage to enable usernames instead of phone numbers!) I would be happy to try "Clear message history", but I'm afraid it would force me to reregister. Will it do so?

There's an even more important issue here. It's not like a few wasted gigs is going to change my life. It's really about what happens to those "deleted" messages. Are you using a per-message unique key that gets overwritten in order to delete each of them permanently, or are they completely overwritten (probably unnecessary), or simply deallocated? Something else? Ideally, they should be gone from my phone in the sense that if you've managed to unlock it and have competent storage browsing tools such as Cellbrite, you're still out of luck.
